### DISCUSSION TITLE
next-auth

### DISCUSSION THREAD

**Discussion by omar2205** [omar2205/next13-i18n-auth-demo#2] [General]
@Rawnly I had to get out of that discussion, as it's only middleware.

When I used your middleware function handlePaths on `/a` or `/whatever` it crashes with `Error: Too many re-renders`.
 
When I try to use `next-auth`, I get hit with `Hydration failed because the initial UI does not match what was rendered on the server.` When I try to use the session.

**Comment by rawnly** [comment]
Can you push the changes? Also have you seen this? https://github.com/nextauthjs/next-auth/issues/5647#issuecomment-1342099364

**Reply by omar2205** [reply]
I'm not using `unstable_getServerSession`. The middleware issue happens with this repo, no need to push any changes. I just did a fresh git clone of this repo and tried to visit `/anything` and it crashed.

**Reply by rawnly** [reply]
just using the snippet i sent you yesterday it runs perfectly for me locally 
![CleanShot 2022-12-09 at 11 56 20@2x](https://user-images.githubusercontent.com/16429579/206687079-3c96cd63-a1d7-4490-820a-51e2e5568cf2.png)

**Reply by omar2205** [reply]
Did you try to visit `/abc`?

**Reply by omar2205** [reply]
I just tried your branch, it still crashes when I tried to visit `/abc`

**Reply by rawnly** [reply]
you're right sorry, anyway i get this 
![CleanShot 2022-12-09 at 13 55 07@2x](https://user-images.githubusercontent.com/16429579/206707286-710221c6-c05c-4ff5-9f6e-15307ed47633.png)
which is not related to the middleware... 

Actually, if you remove the `useTranslation` hook from the page  and layout, it renders with [hydration errors][hydration-error] (probably due to the fact that the entire page is under "use client" )

[hydration-error]: https://nextjs.org/docs/messages/react-hydration-error

Useful links: 
- https://beta.nextjs.org/docs/rendering/fundamentals
- https://nextjs.org/docs/messages/react-hydration-error

**Reply by rawnly** [reply]
![CleanShot 2022-12-09 at 14 05 47@2x](https://user-images.githubusercontent.com/16429579/206709050-e14d3b29-2cad-457a-9ef9-9234ef293c9b.png)

**Reply by omar2205** [reply]
I want it to redirect to `/en/abc`

**Reply by rawnly** [reply]
Oh ok this wasn't clear let me check

**Reply by rawnly** [reply]
@omar2205 somehting like following should work: 

```ts
// inside handlePaths function 
// as second element of the array
  {
    matcher: '/:lang/:path*',
    handler: req => {
      if ( languages.includes(req.params.lang) ) {
        return NextResponse.next()
      }
      
      const lang = getLanguage(req)
      const path = req.params.path.join('/')
      
      return NextResponse.redirect(new URL(`/${lang}/${req.params.lang}/${path}`, req.url))
    }
  }
```

> **NOTE**
> You might need to tweak it as ur needs and update the `/sign-in` and `/login` routes too

**Reply by omar2205** [reply]
I couldn't get this working. I solved it by just sending a 404 response.
```ts
const PUBLIC_FILE = /\.(.*)$/
// ....
  {
    matcher: '/:lang/:path*',
    handler: (req) => {
      if (
        req.nextUrl.pathname.startsWith('/_next') ||
        req.nextUrl.pathname.includes('/api/') ||
        PUBLIC_FILE.test(req.nextUrl.pathname)
      ) {
        return
      }

      const lang = getLanguage(req)
      if (!languages.includes(lang)) {
        return new Response(null, {
          status: 404,
        })
      }
    },
  },
```
